cmd/state.go

```diff
@@ -0,0 +1,167 @@
+	"github.com/pulumi/pulumi/pkg/diag/colors"
+	"github.com/pulumi/pulumi/pkg/resource"
+	"github.com/pulumi/pulumi/pkg/resource/deploy"
+	"github.com/pulumi/pulumi/pkg/resource/edit"
+	"github.com/pulumi/pulumi/pkg/resource/stack"
+	"github.com/pulumi/pulumi/pkg/util/cmdutil"
+	"github.com/spf13/cobra"
+	survey "gopkg.in/AlecAivazis/survey.v1"
+	surveycore "gopkg.in/AlecAivazis/survey.v1/core"
+)
+
+func newStateCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "state",
+		Short: "Edit the current stack's state",
+		Long: `Edit the current stack's state
+		
+Subcommands of this command can be used to surgically edit parts of a stack's state. These can be useful when
+troubleshooting a stack or when performing specific edits that otherwise would require editing the state file by hand.`,
+		Args: cmdutil.NoArgs,
+	}
+
+	cmd.AddCommand(newStateDeleteCommand())
+	cmd.AddCommand(newStateUnprotectCommand())
+	return cmd
+}
+
+// locateStackResource attempts to find a unique resource associated with the given URN in the given snapshot. If the
+// given URN is ambiguous and this is an interactive terminal, it prompts the user to select one of the resources in
+// the list of resources with identical URNs to operate upon.
+func locateStackResource(opts display.Options, snap *deploy.Snapshot, urn resource.URN) (*resource.State, error) {
+	candidateResources := edit.LocateResource(snap, urn)
+	switch {
+	case len(candidateResources) == 0: // resource was not found
+		return nil, nil
+	case len(candidateResources) == 1: // resource was unambiguously found
+		return candidateResources[0], nil
+	}
+
+	// If there exist multiple resources that have the requested URN, prompt the user to select one if we're running
+	// interactively. If we're not, early exit.
+	if !cmdutil.Interactive() {
+		return nil, errors.Errorf("Resource URN %q ambiguously refers to multiple resources", urn)
+	}
+
+	surveycore.DisableColor = true
+	surveycore.QuestionIcon = ""
+	surveycore.SelectFocusIcon = opts.Color.Colorize(colors.BrightGreen + ">" + colors.Reset)
+	prompt := "Multiple resources with the given URN exist, please select the one to edit:"
+	prompt = opts.Color.Colorize(colors.SpecPrompt + prompt + colors.Reset)
+
+	var options []string
+	optionMap := make(map[string]*resource.State)
+	for _, ambiguousResource := range candidateResources {
+		// Prompt the user to select from a list of IDs, since these resources are known to all have the same URN.
+		message := fmt.Sprintf("Resource %q", ambiguousResource.ID)
+		if ambiguousResource.Protect {
+			message += " (Protected)"
+		}
+
+		if ambiguousResource.Delete {
+			message += " (Pending Deletion)"
+		}
+
+		options = append(options, message)
+		optionMap[message] = ambiguousResource
+	}
+
+	var option string
+	if err := survey.AskOne(&survey.Select{
+		Message:  prompt,
+		Options:  options,
+		PageSize: len(options),
+	}, &option, nil); err != nil {
+		return nil, errors.New("no resource selected")
+	}
```

Comment:
should we have an option offering ot cancel?  i'm not sure what we generally do.  i don't necessarily like that hte person could only hit ctrl-c for example to bail out.

Reply:
We don't offer a cancel option for `pulumi new`, which is the other main use of survey selectors.